src/workflow/ArcanistWorkflow.php

```diff
@@ -2068,5 +2100,74 @@ private function getModernCommonDictionary(array $map) {
     return $map;
   }
 
+  // UBER CODE
+  // usso itself is somewhat sluggish, takes 1 second to return cached token
+  const USSO_CACHE_TIMEOUT = 600;
+
+  private static function getUSSOCacheFilename($domain) {
+    return implode(DIRECTORY_SEPARATOR,
+                   array(
+                   sys_get_temp_dir(),
+                         sprintf('usso-token-cache-%s.json', md5($domain)),
+                   ));
+  }
+
+  private function maybeUseUSSOToken($domain) {
+    $cache = self::getUSSOCacheFilename($domain);
+    try {
+      $data = @json_decode(file_get_contents($cache), true);
+      if (is_array($data)) {
+        if (idx($data, 'createdAt', 0) + self::USSO_CACHE_TIMEOUT > time()) {
+          return idx($data, 'token', false);
+        }
+      }
+    } catch (Exception $e) {
+      // it is ok to fail here, we will fallback to `usso`
+    }
+    return false;
+  }
+
+  private function getUSSOTokenForDomain($domain) {
+    // check ussh and if necessary ask to auth
+    list($e, $stdin, $stderr) = exec_manual('ussh');
+    if ($e != 0) {
+      $e = phutil_passthru('ussh');
+      if ($e != 0) {
+        throw new ArcanistUsageException(
+          pht('Looks like you missing `ussh` or it is not working'));
+      }
+    }
+    // try fetching usso token
+    list($e, $stdin, $stderr) = exec_manual('usso -ussh %s', $domain);
+    if ($e != 0) {
+      $e = phutil_passthru('usso -ussh %s', $domain);
+      if ($e != 0) {
+        throw new ArcanistUsageException(
+          pht('Looks like you missing `usso` or it is not working'));
+      }
+    }
+    // try actually fetching token
+    list($e, $stdin, $stderr) = exec_manual('usso -ussh %s -print', $domain);
+    if ($e != 0) {
+      throw new ArcanistUsageException($stdin."\n".$stderr);
+    }
+    $token = trim($stdin);
+    $old = umask(0077);
+    try {
+      $cache = self::getUSSOCacheFilename($domain);
+      file_put_contents($cache,
+                        json_encode(array(
```

Comment:
Indentation is broken.

Reply:
Moved to separate file, fixed idendation